tests/TestCase/Mailer/MailerTest.php

```diff
@@ -0,0 +1,121 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @since         3.1.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\Test\TestCase\Mailer;
+
+use Cake\Mailer\Email;
+use Cake\Mailer\Mailer;
+use Cake\TestSuite\TestCase;
+use TestApp\Mailer\TestMailer;
+
+class MailerTest extends TestCase
+{
+    public function getMockForEmail($methods = [], $args = [])
+    {
+        return $this->getMock('Cake\Mailer\Email', (array)$methods, (array)$args);
+    }
+
+    public function testConstructor()
+    {
+        $mailer = new TestMailer();
+        $this->assertInstanceOf('Cake\Mailer\Email', $mailer->getEmailForAssertion());
+        $this->assertEquals('test', $mailer->layout);
+    }
+
+    public function testReset()
+    {
+        $mailer = new TestMailer();
+        $email = $mailer->getEmailForAssertion();
+
+        $mailer->set(['foo' => 'bar']);
+        $this->assertNotEquals($email->viewVars(), $mailer->reset()->getEmailForAssertion()->viewVars());
+    }
+
+    public function testGetName()
+    {
+        $result = (new TestMailer())->getName();
+        $expected = 'Test';
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testLayout()
+    {
+        $result = (new TestMailer())->layout('foo');
+        $this->assertInstanceOf('TestApp\Mailer\TestMailer', $result);
+        $this->assertEquals('foo', $result->layout);
+    }
+
+    public function testProxies()
+    {
+        $email = $this->getMockForEmail('setHeaders');
+        $email->expects($this->once())
+            ->method('setHeaders')
+            ->with([]);
+        $result = (new TestMailer($email))->setHeaders([]);
+        $this->assertInstanceOf('TestApp\Mailer\TestMailer', $result);
+
+        $email = $this->getMockForEmail('addHeaders');
+        $email->expects($this->once())
+            ->method('addHeaders')
+            ->with([]);
+        $result = (new TestMailer($email))->addHeaders([]);
+        $this->assertInstanceOf('TestApp\Mailer\TestMailer', $result);
+
+        $email = $this->getMockForEmail('attachments');
+        $email->expects($this->once())
+            ->method('attachments')
+            ->with([]);
```

Comment:
It might be good to check these with non empty arguments to ensure arguments are passed correctly.